### PR TITLE
User/v royad/disable anonymous access on containers

### PR DESCRIPTION
**Pull request by RohitYadav-msft**

### CODE REVIEW

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 1]
```diff
@@ -0,0 +1,382 @@
+﻿﻿function Pre_requisites
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1:1]: Added

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        Install-Module -Name Az.Resources -Scope CurrentUser
+    }
+    else
+    {
+        Write-Host "Az.Resources module is available." -ForegroundColor Green
+    }
+}
+
+
+function Remediate-DisableAnonymousAccessOnContainers
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1]: Added function SYNOPSIS

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+
+    Write-Host "======================================================"
+    Write-Host "Starting to disable anonymous access on containers of storage account for subscription."
+    Write-Host "------------------------------------------------------"
+
+    if($PerformPreReqCheck)
+    {
+       Write-Host "Checking for pre-requisites..."
+       Pre_requisites
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1]: Yes we should return if any exception occurred while performing pre_requisites check

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    }
+
+    # Setting context for current subscription.
+    $currentSub = Set-AzContext -SubscriptionId $SubscriptionId
+
+    Write-Host "Metadata Details: `n SubscriptionId: [$($SubscriptionId)] `n AccountName: [$($currentSub.Account.Id)] `n AccountType: [$($currentSub.Account.Type)]"
+    Write-Host "------------------------------------------------------"
+    Write-Host "Starting with Subscription [$($SubscriptionId)]..."
+
+    # Safe Check: Checking whether the current account is of type User and also grant the current user as UAA for the sub to support fallback
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1]: Removed "and also grant the current user as UAA for the sub to support fallback",  it is not valid in this case

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    else
+    {
+        if (-not (Test-Path -Path $Path))
+        {
+            Write-Host "Error: Control file path is not found." -ForegroundColor Red
+            exit;        
+        }
+
+        # Fetching failed controls details from given json file.
+        $ControlIds = "Azure_Storage_AuthN_Dont_Allow_Anonymous"
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1]: Updated

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        $resourceDetails = $controls | Where-Object { $ControlIds -eq $_.ControlId};
+
+        if(($resourceDetails | Measure-Object).Count -eq 0 -and ($resourceDetails.ResourceDetails | Measure-Object).Count -eq 0)
+        {
+            Write-Host "No resource found in input json file for remedition." -ForegroundColor Red
+            exit;
+        }
+        $resourceDetails.ResourceDetails | ForEach-Object { 
+                            try{
+                                $resourceContext += Get-AzStorageAccount -Name $_.ResourceName -ResourceGroupName $_.ResourceGroupName
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1]: Using resource id not able to fetch storage account detail.
It show parameter not matched...

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+                            }
+                            catch
+                            {
+                                Write-Host "Valid resource group and resource name not found in input json file. ErrorMessage [$($_)]" -ForegroundColor Red  
+                            }
+                        }
+    }
+    
+    # Performing remediation
+    try
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1]: Added

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+        {
+            $ContainersWithAnonymousAccessOnStorage = @();
+            $ContainersWithDisableAnonymousAccessOnStorage = @();
+            $resourceContext | ForEach-Object{
+                $flag = $true
+                $allContainers = @();
+                $containersWithAnonymousAccess = @();
+                $anonymousAccessContainersNameAndPublicAccess = @();
+                $context = $_.context;
+                $allContainers += Get-AzStorageContainer -Context $context -ErrorAction Stop
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1]: Updated control logic to support both type of remediation "DisableAnonymousAccessOnContainers", "EnableAllowBlobPublicAccessOnStorage" , User can select remediation type to fix issue using remediation command.

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1 [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+                if($allContainers.Count -ne 0)
+                {
+                    $containersWithAnonymousAccess += $allContainers | Where-Object { $_.PublicAccess -ne "Off"}
```
[the remark this answers is not in the document]
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-DisableAnonymousAccessOnContainers.ps1]: Updated logic

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-AnonymousAccessOnContainers.ps1 [hunk 10]
```diff
@@ -0,0 +1,639 @@
...
+    Write-Host "Metadata Details: `n SubscriptionId: [$($SubscriptionId)] `n AccountName: [$($currentSub.Account.Id)] `n AccountType: [$($currentSub.Account.Type)]"
+    Write-Host $([Constants]::SingleDashLine)  
+    Write-Host "Starting with Subscription [$($SubscriptionId)]..."
+
+    Write-Host "`n"
+    Write-Host "*** To perform remediation for disabling anonymous access on containers user must have atleast contributor access on storage account of Subscription: [$($SubscriptionId)] ***" -ForegroundColor $([Constants]::MessageType.Warning)
+    Write-Host "`n" 
+
+    Write-Host "Validating whether the current user [$($currentSub.Account.Id)] have the required permissions to run the script for Subscription [$($SubscriptionId)]..."
+    Write-Host "`n"
```
Comment (TarunKrShukla) [inline 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-AnonymousAccessOnContainers.ps1:131]: At which step we are actually for permissions.

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-AnonymousAccessOnContainers.ps1 [hunk 11]
```diff
@@ -0,0 +1,639 @@
...
+        "DisableAllowBlobPublicAccessOnStorage"
+        {
+            try
+            {
+                if($totalStorageAccount -gt 0)
+                {
+                    $stgWithEnableAllowBlobPublicAccess = @()
+                    $stgWithDisableAllowBlobPublicAccess = @()
+                    $resourceContext | ForEach-Object {
+                        $stg = Get-AzStorageAccount -ResourceGroupName $_.ResourceGroupName -StorageAccountName $_.StorageAccountName
```
Comment (TarunKrShukla) [inline 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-AnonymousAccessOnContainers.ps1:195]: $resourceContext already has output of Get-AzStorageAccount, do we need to call Get-AzStorageAccount again here?

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-AnonymousAccessOnContainers.ps1 (hunk 10 above)
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-AnonymousAccessOnContainers.ps1:131]: Updated comments

File: 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-AnonymousAccessOnContainers.ps1 (hunk 11 above)
Comment (RohitYadav-msft) [reply 13-Tenant Security Solution Setup/Scripts/RemediationScripts/Remediate-AnonymousAccessOnContainers.ps1]: Updated script, no need to call 'Get-AzStorageAccount' command, we have all required data to check further condition.

**Review by asimks — APPROVED** [review]

**Review by TarunKrShukla — APPROVED** [review]

**Pull request merged**